crates/starfish/core/src/network/mod.rs

```diff
@@ -286,7 +286,84 @@ impl TryFrom<SerializedBlock> for SerializedHeaderAndTransactions {
     }
 }
 
+#[derive(Clone, Debug, PartialEq)]
+pub(crate) struct BlockBundle {
+    pub(crate) verified_block: VerifiedBlock,
+    pub(crate) verified_headers: Vec<VerifiedBlockHeader>,
+}
+
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+pub(crate) struct SerializedBlockAndHeaders {
+    pub(crate) serialized_block: Bytes,
+    pub(crate) serialized_headers: Vec<Bytes>,
+}
+
 #[derive(Clone, Debug, PartialEq)]
 pub(crate) struct SerializedBlockBundle {
     pub(crate) serialized_block_bundle: Bytes,
 }
+
+impl TryFrom<VerifiedBlock> for SerializedBlockAndHeaders {
+    type Error = ConsensusError;
+    fn try_from(verified_block: VerifiedBlock) -> ConsensusResult<Self> {
+        let (serialized_block_header, serialized_transactions) = verified_block.serialized();
+        let serialized_header_and_transactions = SerializedHeaderAndTransactions {
+            serialized_block_header: serialized_block_header.clone(),
+            serialized_transactions: serialized_transactions.clone(),
+        };
+        let bytes = bcs::to_bytes(&serialized_header_and_transactions)
+            .map_err(ConsensusError::SerializationFailure)?;
+        Ok(Self {
+            serialized_block: Bytes::from(bytes),
+            serialized_headers: vec![],
+        })
+    }
+}
+
+impl TryFrom<BlockBundle> for SerializedBlockAndHeaders {
+    type Error = ConsensusError;
+    fn try_from(block_bundle: BlockBundle) -> ConsensusResult<Self> {
+        let (serialized_block_header, serialized_transactions) =
+            block_bundle.verified_block.serialized();
+        let serialized_header_and_transactions = SerializedHeaderAndTransactions {
+            serialized_block_header: serialized_block_header.clone(),
+            serialized_transactions: serialized_transactions.clone(),
+        };
+        let bytes = bcs::to_bytes(&serialized_header_and_transactions)
+            .map_err(ConsensusError::SerializationFailure)?;
+        let mut serialized_block_headers = vec![];
+        for block_header in block_bundle.verified_headers.into_iter() {
+            serialized_block_headers.push(block_header.serialized().clone());
+        }
+
+        Ok(Self {
+            serialized_block: Bytes::from(bytes),
+            serialized_headers: serialized_block_headers,
+        })
+    }
+}
+
+impl TryFrom<SerializedBlockAndHeaders> for SerializedBlockBundle {
+    type Error = ConsensusError;
+    fn try_from(serialized_block_and_headers: SerializedBlockAndHeaders) -> ConsensusResult<Self> {
+        let bytes = bcs::to_bytes(&serialized_block_and_headers)
+            .map_err(ConsensusError::SerializationFailure)?;
+        Ok(Self {
+            serialized_block_bundle: Bytes::from(bytes),
+        })
+    }
+}
+
+impl TryFrom<VerifiedBlock> for SerializedBlockBundle {
```

Comment:
I don't think we should be creating VerifiedBlock without actually verifying the contents of the header and transactions. In `handle_subscribed_block` method we receive `SerializedBlock` instance and only create `VerifiedBlock` after successful verification. On the serialization part it's fine, but we cannot optimistically assume that what we receive is valid and correct.

Reply:
ah, ok. My bad, understood it the wrong way